x-pack/plugins/ml/public/application/components/data_grid/data_grid.tsx

```diff
@@ -50,132 +51,152 @@ function isWithHeader(arg: any): arg is PropsWithHeader {
-            </EuiCodeBlock>
+            <p>{noDataMessage}</p>
           </EuiCallOut>
-          <EuiSpacer size="m" />
         </div>
-      )}
-      <EuiDataGrid
-        aria-label={isWithHeader(props) ? props.title : ''}
-        columns={columns}
-        columnVisibility={{ visibleColumns, setVisibleColumns }}
-        gridStyle={euiDataGridStyle}
-        rowCount={rowCount}
-        renderCellValue={renderCellValue}
-        sorting={{ columns: sortingColumns, onSort }}
-        toolbarVisibility={euiDataGridToolbarSettings}
-        pagination={{
-          ...pagination,
-          pageSizeOptions: [5, 10, 25],
-          onChangeItemsPerPage,
-          onChangePage,
-        }}
-      />
-    </div>
-  );
-};
+      );
+    }
+
+    return (
+      <div data-test-subj={`${dataTestSubj} ${status === INDEX_STATUS.ERROR ? 'error' : 'loaded'}`}>
+        {isWithHeader(props) && (
+          <EuiFlexGroup alignItems="center" justifyContent="spaceBetween">
+            <EuiFlexItem>
+              <DataGridTitle title={props.title} />
+            </EuiFlexItem>
+            <EuiFlexItem grow={false}>
+              <EuiCopy
+                beforeMessage={props.copyToClipboardDescription}
+                textToCopy={props.copyToClipboard}
+              >
+                {(copy: () => void) => (
+                  <EuiButtonIcon
+                    onClick={copy}
+                    iconType="copyClipboard"
+                    aria-label={props.copyToClipboardDescription}
+                  />
+                )}
+              </EuiCopy>
+            </EuiFlexItem>
+          </EuiFlexGroup>
+        )}
+        {status === INDEX_STATUS.ERROR && (
+          <div data-test-subj={`${dataTestSubj} error`}>
+            <EuiCallOut
+              title={i18n.translate('xpack.ml.dataGrid.indexDataError', {
+                defaultMessage: 'An error occurred loading the index data.',
+              })}
+              color="danger"
+              iconType="cross"
+            >
+              <EuiCodeBlock language="json" fontSize="s" paddingSize="s" isCopyable>
+                {errorMessage}
+              </EuiCodeBlock>
+            </EuiCallOut>
+            <EuiSpacer size="m" />
+          </div>
+        )}
+        <EuiDataGrid
+          aria-label={isWithHeader(props) ? props.title : ''}
+          columns={columns}
+          columnVisibility={{ visibleColumns, setVisibleColumns }}
+          gridStyle={euiDataGridStyle}
+          rowCount={rowCount}
+          renderCellValue={renderCellValue}
+          sorting={{ columns: sortingColumns, onSort }}
+          toolbarVisibility={euiDataGridToolbarSettings}
+          pagination={{
+            ...pagination,
+            pageSizeOptions: [5, 10, 25],
+            onChangeItemsPerPage,
+            onChangePage,
+          }}
+        />
+      </div>
+    );
+  },
+  (prevProps, nextProps) => isEqual(pickProps(prevProps), pickProps(nextProps))
```

Comment:
I presume it won't pay off with significant performance benefits. Instead the parent component that provides these props should update them only when necessary, hence default shallow comparison is this component will suffice

Reply:
At least in my testing it was a quick win. We use the component in 5 different places and I didn't want to touch the outer code for every place we use it. Without the `React.memo` the render of `EuiDataGrid` would trigger a lot more and it's a really slow component. We can look into further improvements but let's do that in follow-ups since this would also touch the data frame analytics results pages.